# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/url-classifier/UrlClassifierFeatureFactory.cpp
# Commit: d393067941bd
# Full Hash: d393067941bd69dfa7eff7790e19a0a0162076cd
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2022-07-20 09:59:35
# Regressor Bug: 1773701
# File Overlap Count: 1
# Description:
#   Bug 1773701 - Part 2: Implement the email tracking data collection feature. r=dimi
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D151522
# ==============================================================================

diff -r f3c636e27b0b -r d393067941bd netwerk/url-classifier/UrlClassifierFeatureFactory.cpp
--- a/netwerk/url-classifier/UrlClassifierFeatureFactory.cpp	Tue Jul 19 20:04:24 2022 +0000
+++ b/netwerk/url-classifier/UrlClassifierFeatureFactory.cpp	Tue Jul 19 20:04:24 2022 +0000
@@ -9,6 +9,7 @@
 // List of Features
 #include "UrlClassifierFeatureCryptominingAnnotation.h"
 #include "UrlClassifierFeatureCryptominingProtection.h"
+#include "UrlClassifierFeatureEmailTrackingDataCollection.h"
 #include "UrlClassifierFeatureEmailTrackingProtection.h"
 #include "UrlClassifierFeatureFingerprintingAnnotation.h"
 #include "UrlClassifierFeatureFingerprintingProtection.h"
@@ -35,6 +36,7 @@
 
   UrlClassifierFeatureCryptominingAnnotation::MaybeShutdown();
   UrlClassifierFeatureCryptominingProtection::MaybeShutdown();
+  UrlClassifierFeatureEmailTrackingDataCollection::MaybeShutdown();
   UrlClassifierFeatureEmailTrackingProtection::MaybeShutdown();
   UrlClassifierFeatureFingerprintingAnnotation::MaybeShutdown();
   UrlClassifierFeatureFingerprintingProtection::MaybeShutdown();
@@ -60,6 +62,16 @@
   // feature order, and this could produce different results with a different
   // feature ordering.
 
+  // Email Tracking Data Collection
+  // This needs to be run before other features so that other blocking features
+  // won't stop us to collect data for email trackers. Note that this feature
+  // is not a blocking feature.
+  feature =
+      UrlClassifierFeatureEmailTrackingDataCollection::MaybeCreate(aChannel);
+  if (feature) {
+    aFeatures.AppendElement(feature);
+  }
+
   // Email Tracking Protection
   feature = UrlClassifierFeatureEmailTrackingProtection::MaybeCreate(aChannel);
   if (feature) {
@@ -148,6 +160,13 @@
     return feature.forget();
   }
 
+  // Email Tracking Data Collection
+  feature =
+      UrlClassifierFeatureEmailTrackingDataCollection::GetIfNameMatches(aName);
+  if (feature) {
+    return feature.forget();
+  }
+
   // Email Tracking Protection
   feature =
       UrlClassifierFeatureEmailTrackingProtection::GetIfNameMatches(aName);
@@ -230,6 +249,12 @@
     aArray.AppendElement(name);
   }
 
+  // Email Tracking Data Collection
+  name.Assign(UrlClassifierFeatureEmailTrackingDataCollection::Name());
+  if (!name.IsEmpty()) {
+    aArray.AppendElement(name);
+  }
+
   // Email Tracking Protection
   name.Assign(UrlClassifierFeatureEmailTrackingProtection::Name());
   if (!name.IsEmpty()) {